Introduce hacking guide how to build and test features for CCCMO (#50)
openshift/cluster-cloud-controller-manager-operator: docs/dev/hacking-guide.md

## How to run a component locally for testing

Prerequisites:

```bash
git checkout github.com/openshit/$repository_name
cd $repository_name
```

Make sure your $KUBECONFIG is set properly, because it will be used to interact with your cluster.

First step is to scale down the cluster version operator.

“At the heart of OpenShift is an operator called the Cluster Version Operator. This operator watches the deployments and images related to the core OpenShift services, and will prevent a user from changing these details. If I want to replace the core OpenShift services I will need to scale this operator down.”

```bash
oc scale --replicas=0 deployment/cluster-version-operator -n openshift-cluster-version
```

Second step is scaling down the CCCMO. The operator watches for deployed resources to be running and will prevent them from scaling down.

```bash
oc scale deployment/cluster-cloud-controller-manager -n openshift-cloud-controller-manager-operator --replicas=0
```

Finally, once all has been scaled down you can compile and run the controller.

You can copy or edit `hack/example-images.json` with the provider image you want to run.

```bash
make build

>  ./bin/cluster-cloud-controller-manager --images-json=hack/example-images.json

Author: It is currently called `cluster-controller-manager-operator` This value is not exposed anywhere, and in the context of project event something like `cloud-manager-operator` could work. Or `cluster-cloud-controller`, or even `cluster-cloud-manager`.

```

## How to build the operator in a container for remote testing

Prerequisites:

```bash
git checkout github.com/openshit/$repository_name
cd $repository_name
```

First you’ll need to build an image with your changes.

```bash

> IMG="quay.io/your-repo/cluster-cloud-controller-manager-operator:latest" make images
> docker push quay.io/your-repo/cluster-cloud-controller-manager-operator:latest

Reviewer: I'd suggest instead of using the latest tag, we suggest to users to tag their work with something meaningful, eg their branch name?
```suggestion
IMG="quay.io/your-repo/cluster-cloud-controller-manager-operator:<branch-name>" make images
docker push quay.io/your-repo/cluster-cloud-controller-manager-operator:<branch-name>
```

```

Second step, similar to running the component locally we need to scale down CVO.

```bash
oc scale --replicas=0 deployment/cluster-version-operator -n openshift-cluster-version
```

> Next, edit machine-api-operator-images ConfigMap and place a link to your image there.

Reviewer: MAO? I don't think that's right 😅  Let's include the namespace for the configmap too please

Reviewer 2: we should put the names that people are expected to lookup in the cluster in block quotes too, eg
```suggestion
Next, edit `machine-api-operator-images` ConfigMap and place a link to your image there.
```

> The clsuter-cloud-controller-manager-operator project contains all the resources and components that are used by the CCCMO. There is a ConfigMap named cloud-controller-manager-images. This ConfigMap contains references to all the images used by the CCCMO, it uses these to deploy the CCM and ensure they are the proper images. The ConfigMap looks something like this:

Reviewer: Suggested typo and a bit more context for the last bit of this
```suggestion
The cluster-cloud-controller-manager-operator project contains all the resources and components that are used by the CCCMO.
There is a ConfigMap named cloud-controller-manager-images.
This ConfigMap contains references to all the images used by the CCCMO.
It uses these images to deploy the CCM to various platforms and ensure that the running images are from the correct release payload.

The ConfigMap looks something like this:
```

Reviewer 2: similar to above, we should wrap exact names that would be seen in the cluster with block quotes. in this case the operator project name, and configmap

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cloud-controller-manager-images
  namespace: openshift-cloud-controller-manager-operator
data:
  images.json: |
    {
      "cloudControllerManagerAWS": "registry.ci.openshift.org/openshift:aws-cloud-controller-manager",
      "cloudControllerManagerOpenStack": "registry.ci.openshift.org/openshift:openstack-cloud-controller-manager"
    }
```

```bash
oc edit configmap/cloud-controller-manager-images -n openshift-cloud-controller-manager-operator
```

> With the new image information loaded into the ConfigMap, the next thing you might do is replace the CCCMO. This operator controls how the specific cloud controllers are deployed and coordinated. You only change this component if there is something you are testing.

Reviewer: Do we actually need to update the images? CCCMO doesn't have its own image in the configmap, and it doesn't use this anywhere, so if we are updating CCCMO, I don't think we need the configmap edits do we?

> The easiest way to change this operator is to change the image reference in the deployment. The commands you can use:

Reviewer 2: minor nit
```suggestion
The easiest way to change this operator is to change the image reference in the deployment. You can use the following command to edit image references:
```

```bash
oc edit deployment/cluster-cloud-controller-manager -n openshift-cloud-controller-manager-operator
```

After changing the download reference in the images ConfigMap an easy way to swap out the controller is to let the CCCMO do it for you. You can delete the deployment associated with the cloud provider CCM and then the CCCMO will create a new one for you, like this:

```bash
oc delete deployment/aws-cloud-controller-manager -n openshift-cloud-controller-manager

> ```

Reviewer: I'm a bit confused, can we add more details to the beginning of the sections in this doc to explain what we are achieving here, I thought we were hacking on CCCMO but this suggests we are hacking on CCM itself?

Author: It is a dev guide. Hacking is a nature of it, and it makes sense to name it so, as longer term this would become a nice-to-have dev documentation for all and everything.

Reviewer 2: imo, seems like both activities are inter-linked. i think as we grow this into the devel style doc that Danil is talking about it will be come more important to create separate CCM and CCCMO sections, but for now this is reading ok to me.

To achieve this you need to build custom release payload. To do that you should use `oc adm release new`.

Here is an example how to build a release with changes in CCCMO, KCMO and MCO at the same time. Precondition for this step - you have to build each operator image individually and push to a public registry.

```bash
oc adm release new --server=https://api.ci.l2s4.p1.openshiftapps.com:6443 \ # CI server to pull existing release payload from
  --from-release=registry.ci.openshift.org/ocp/release:4.8.0-0.ci-2021-05-24-100659 \ # Your release tag to pull the rest of the images from
  --registry-config=~/pull-secret \ # Your pull-secret with permissions to read from CI server
  --to-image=quay.io/dgrigore/release:latest \ # Where the release image will be pushed
  # Now your image overrides for image tags
    machine-config-operator=quay.io/dgrigore/machine-config-operator:ccm \
    cluster-kube-controller-manager-operator=quay.io/dgrigore/cluster-kube-controller-manager-operator:azure-bootstrap \
    cluster-cloud-controller-manager-operator=quay.io/dgrigore/cluster-cloud-controller-manager-operator:azure-bootstrap
```

> At the current stage of development operator is only present in CI builds, but is not present in nightlies and is not in release. You could pick one of available CI releases from the list here: https://amd64.ocp.releases.ci.openshift.org/#4.8.0-0.ci.

Reviewer: Readability improvements
```suggestion
At the current stage of development, the operator is only present in CI builds. 
It is not present in nightly or release builds.
You can pick one of available CI releases from the list here: https://amd64.ocp.releases.ci.openshift.org/#4.8.0-0.ci.
```

> You might have to pick a fresh one once in couple of days, as they are constatntly being removed, and run `oc adm release new` once again.

Reviewer: Suggested readability improvement
```suggestion
CI builds are pruned regularly and as such, you will have to create a new release every few days with an updated base release. Follow the process above and run `oc adm release new` once again.
```

Tags for image overrides are coming from the operator manifests itself. The image tags could be found in `manifests/image-references`. Example might be KCMO image references: https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/manifests/image-references

> In this scenario, you build KCMO operator image with `make image`, push it somewhere like `quay.io/dgrigore/cluster-kube-controller-manager-operator:azure-bootstrap` and then specify the override like: 
> 
> ```bash
> ...
> cluster-kube-controller-manager-operator=quay.io/dgrigore/cluster-kube-controller-manager-operator:azure-bootstrap 
> ...
> ```

Reviewer: Might be worth making these examples more generic, so they don't directly relate to your repo
```suggestion
In this scenario, you build KCMO operator image with `make image`, push it somewhere like `quay.io/<my-repo>/cluster-kube-controller-manager-operator:<my-branch>` and then specify the override like: 

```bash
...
cluster-kube-controller-manager-operator=quay.io/<my-repo>/cluster-kube-controller-manager-operator:<my-branch>
...
```
```

Reviewer 2: ++ i was thinking something similar

By specifying the `CustomNoUpgrade` feature gate in the cluster with `ExternalCloudProvider` you are enabling CCCMO logic for providers which are currently in Technical Preview state.

```bash
# This step would stop after creation of ./manifests folder. All resources placed there will be created in the cluster
# during bootstrap and would override the default state of the resource in the cluster.
OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE=quay.io/dgrigore/release:latest openshift-install create manifests

cat <<EOF > manifests/manifest_feature_gate.yaml
apiVersion: config.openshift.io/v1
kind: FeatureGate
metadata:
  annotations:
    include.release.openshift.io/self-managed-high-availability: "true"
    include.release.openshift.io/single-node-developer: "true"
    release.openshift.io/create-only: "true"
  name: cluster
spec:
  customNoUpgrade:
    enabled:
    - ExternalCloudProvider
    - CSIMigrationAWS
    - CSIMigrationOpenStack
  featureSet: CustomNoUpgrade
EOF

# Now you could create cluster with your custom release
OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE=quay.io/dgrigore/release:latest openshift-install create cluster
```

> If you additionally whant to include installer changes, you need to build installer binary from https://github.com/openshift/installer

Reviewer: ```suggestion
If you additionally want to include installer changes, you need to build installer binary from https://github.com/openshift/installer
```

# Cluster cloud controller manage operator - Hacking guide

## How to run a component locally for testing

Prerequisites:

```bash
git checkout github.com/openshit/$repository_name
cd $repository_name
```

Make sure your `$KUBECONFIG` is set properly, because it will be used to interact with your cluster.

First step is to scale down the cluster version operator.

> “At the heart of OpenShift is an operator called the Cluster Version Operator. This operator watches the deployments and images related to the core OpenShift services, and will prevent a user from changing these details. If I want to replace the core OpenShift services I will need to scale this operator down.”

```bash
oc scale --replicas=0 deployment/cluster-version-operator -n openshift-cluster-version
```

Second step is scaling down the CCCMO. The operator watches for deployed resources to be running and will prevent them from scaling down.

```bash
oc scale deployment/cluster-cloud-controller-manager -n openshift-cloud-controller-manager-operator --replicas=0
```

> Finally, once all has been scaled down you can compile and run the controller.
> 
> You can copy or edit `hack/example-images.json` with the provider image you want to run.

Reviewer 2: i think you should combine these two sentences, it will make more sense. 
```suggestion
Finally, once all has been scaled down you can compile and run the controller. If you need to run additional images you can copy or edit `hack/example-images.json` to add the images you want to run.
```

Author: Amended it slightly, so this sentence makes more sense.

```

With the new image information loaded into the ConfigMap, the next thing you might do is replace the CCCMO. This operator controls how the specific cloud controllers are deployed and coordinated. You only change this component if there is something you are testing.

The easiest way to change this operator is to change the image reference in the deployment. You can use the following command to edit image references or other things in the deployment:

```bash
oc edit deployment/cluster-cloud-controller-manager -n openshift-cloud-controller-manager-operator
```

If you want images ConfigMap changes to reconcile you need to cause creation of new replicaSet for CCM, which will ensure the new content of ConfigMap is mounted.

After changing the download reference in the images ConfigMap an easy way to swap out the controller is to let the CCCMO do it for you. You can delete the deployment associated with the cloud provider CCM and then the CCCMO will create a new one for you, like this:

```bash
oc delete deployment/aws-cloud-controller-manager -n openshift-cloud-controller-manager
```

## How to build operator in combination with other operators and build a release image

> At the current stage of development it is frequent that for testing a feature, another openshift operator change should be in place to make the feature work in CCCMO.

Reviewer 2: this could probably use a little more explanation, i'm not sure which operator change you are talking about when you mention "another openshift operator change should be in place". 

or is this just saying that _in general_ you might need to make other changes?

Author: In general. This operator lives between multiple other components, and so the changes in one place create changes in another.
